### ISSUE TITLE
[Feature Request] Option to keep toolbars inlined with view/editors tabs

### ISSUE THREAD

**Issue by MarioMarinatoSTech** [eclipse-platform/eclipse.platform.ui#1160]
I work on a team that develops an application based on Eclipse RCP.  

As requested by our customers, we implemented some visual modifications regarding the way the view toolbar items behave when there is not enough space for them all to be drawn on the same line. 

Eclipse's default behavior is to move the toolbar and view menu to the bottom of the tab and add more lines to hold the items, as necessary. 

Here are some images of Eclipse's default behavior, using a screen from our application as an example: 

Window with space for all items: 
![img1](https://github.com/eclipse-platform/eclipse.platform.ui/assets/113549558/48a6cc70-59d7-4d87-ad49-4c713f629822)

Window without enough space, creating a line below: 
![img2](https://github.com/eclipse-platform/eclipse.platform.ui/assets/113549558/f3afcff0-5495-48d1-8529-2287a0bcbb30)

Window without enough space, creating more lines: 
![img3](https://github.com/eclipse-platform/eclipse.platform.ui/assets/113549558/fba0c036-25b9-4117-bed5-31428b72e45b)

We implemented was an option for the toolbars to be kept in a single line, next to the tabs, as shown in the next images. 

Window without enough space. Therefore, each hidden item generates an entry in the “chevron” menu: 
![img4](https://github.com/eclipse-platform/eclipse.platform.ui/assets/113549558/7efc2543-c1e7-4dc2-8abf-28d571a80f3e)

Window without enough space. In this case, all toolbar items went to the chevron menu. Note that the view menu also remained on the same line, next to Minimize button: 
![img5](https://github.com/eclipse-platform/eclipse.platform.ui/assets/113549558/07717a64-7959-41a6-961a-1cdac004f54e)

Using the Eclipse 4 framework, we made a specialization of the `StackRenderer` class, responsible for managing the area called “top right” of the `CTabFolder`, where the view menu and the toolbar are located.  In this specialization, we reimplemented the `addTopRight` and `adjustTopRight` methods, changing the parent of the `ToolBar`, which is the toolbar component, to be a `CoolBar`, so to get the “chevron” behavior.  Other minor modifications were made to these methods and the `ToolBar` class. 

As a way to make Eclipse use this class, we created a Model Fragment with an Addon, which is responsible for overriding the "rendererFactoryUri", and, therefore, when any `MPartStack` element needs to be created, our class it is used. 

In several parts of this process, we need to access restricted parts of the Eclipse code. 

We think this setting could be useful on Eclipse platform itself.  Unfortunately, due to lack of time, we cannot contribute directly to the project ourselves.  If there is interest in implementing this feature, we're happy to share our code.

**Comment by vogella** [comment]
Looks way nicer. I personally think this should replace our current implementation.

> Unfortunately, due to lack of time, we cannot contribute directly to the project ourselves. If there is interest in implementing this feature, we're happy to share our code.

That should be really easy these days, just fork platform.ui, push to it and  create a PR. Typically this takes not very long.

**Comment by iloveeclipse** [comment]
> I personally think this should replace our current implementation.

While compact mode itself looks clean, it has a major usability drawback. For example, in this screenshot the buttons "Start" and "Stop" disappear and user would need more time/UI interactions to open menu and trigger these operations.

While for some views this could be irrelevant, others may want some buttons to permanently shown. In case this feature would be added, it shouldn't be enabled by default or it should be possible for views / users to switch it off.

**Comment by vogella** [comment]
@MarioMarinatoSTech please paste your code into this issue in case you do not have the time to create a PR. I would like to use it. :-)

**Comment by MarioMarinatoSTech** [comment]
Ok, a co-worker and I talked and we decided to give it a try.  :-)

We thought of doing the following:
 - Add a new setting on Preferences > General > Appearance ( [  ] Inline toolbars ) 
 - By default, this option will be off
 - When the user enables it, the toolbars start to behave as we described above.

Do you think the text and location are fine?

**Comment by vogella** [comment]
Sounds perfect to me

**Comment by laeubi** [comment]
Why not doing it the E4/RCP way and having an Addon that acts on a `TAG`?
The one can configure this on a per view basis what seems most useful as there might be case where one wants this and case where not in the same application...

**Comment by vogella** [comment]
> Why not doing it the E4/RCP way and having an Addon that acts on a `TAG`? The one can configure this on a per view basis what seems most useful as there might be case where one wants this and case where not in the same application...

Lets try to start simple and see if this needs extensions later. Having this in the IDE as an option for all views is already a big step for usability in my opinion

**Comment by laeubi** [comment]
> > Why not doing it the E4/RCP way and having an Addon that acts on a `TAG`? The one can configure this on a per view basis what seems most useful as there might be case where one wants this and case where not in the same application...
> 
> Lets try to start simple and see if this needs extensions later. Having this in the IDE as an option for all views is already a big step for usability in my opinion

I don't think this is actually "simpler" especially as (UI) Preferences are only available on compatE4 and not plain RCP apps. Of course one can have a Preference in Eclipse(!) UI that then set this on all parts, but I think a tag is much more flexible and suitable for such feature.

**Comment by vogella** [comment]
@MarioMarinatoSTech any update on the PR?

**Comment by MarioMarinatoSTech** [comment]
Hey, @vogella, we're going to get into it this week.  I'm confident that it'll be done before the weekend.
